File: frontend/src/api/opportunities.ts
/**
 * Phase14-G1.1 — Opportunity user action API client.
 */

export const OPPORTUNITY_ACTION = {
  VIEW: 'VIEW',
  WATCH: 'WATCH',
  IGNORE: 'IGNORE',
} as const

export type OpportunityActionType = (typeof OPPORTUNITY_ACTION)[keyof typeof OPPORTUNITY_ACTION]

export type OpportunityUserActionSummary = {
  id: string
  action: string
  created_at: string
  created_by?: string
  opportunity_id: string
}

export type OpportunityPoolEntry = {
  opportunity_id: string
  scan_batch_key: string
  stock_code: string
  secucode: string
  stock_name?: string
  signal_tag?: string
  signal_time?: string
  latest_user_action?: OpportunityUserActionSummary | null
}

export type SaveOpportunityActionRequest = {
  accountId?: number
  scanBatchKey: string
  opportunityId?: string
  stockCode?: string
  secucode?: string
  signalTime?: string
  signalTag?: string
  action: OpportunityActionType
  createdBy?: string
}

export type OpportunityListQuery = {
  tradeDate?: string
  session?: string
  strategyId?: string
  accountId?: number
  includeUserAction?: boolean
}

function mapActionSummary(raw: Record<string, unknown> | null | undefined) {
  if (!raw?.action) return null
  return {
    id: String(raw.id || ''),
    action: String(raw.action || ''),
    created_at: String(raw.created_at || raw.createdAt || ''),
    created_by: raw.created_by ? String(raw.created_by) : undefined,
    opportunity_id: String(raw.opportunity_id || raw.opportunityId || ''),
  } satisfies OpportunityUserActionSummary
}

function mapEntry(raw: Record<string, unknown>): OpportunityPoolEntry {
  return {
    opportunity_id: String(raw.opportunity_id || raw.opportunityId || ''),
    scan_batch_key: String(raw.scan_batch_key || raw.scanBatchKey || ''),
    stock_code: String(raw.stock_code || raw.stockCode || ''),
    secucode: String(raw.secucode || raw.SECUCODE || ''),
    stock_name: raw.stock_name ? String(raw.stock_name) : undefined,
    signal_tag: raw.signal_tag ? String(raw.signal_tag) : undefined,
    signal_time: raw.signal_time ? String(raw.signal_time) : undefined,
    latest_user_action: mapActionSummary(
      (raw.latest_user_action || raw.latestUserAction) as Record<string, unknown>,
    ),
  }
}

/** POST /api/opportunities/action */
export async function postOpportunityAction(req: SaveOpportunityActionRequest) {
  const res = await fetch('/api/opportunities/action', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      account_id: req.accountId,
      scan_batch_key: req.scanBatchKey,
      opportunity_id: req.opportunityId,
      stock_code: req.stockCode,
      secucode: req.secucode,
      signal_time: req.signalTime,
      signal_tag: req.signalTag,
      action: req.action,
      created_by: req.createdBy || 'ui:opportunity-list',
    }),
  })
  const body = (await res.json()) as Record<string, unknown>
  if (!res.ok || !body?.ok) {
    throw new Error(String(body?.message || `HTTP ${res.status}`))
  }
  return body.action as Record<string, unknown>
}

/** GET /api/opportunities/list */
export async function fetchOpportunityList(q: OpportunityListQuery = {}) {
  const params = new URLSearchParams()
  if (q.tradeDate?.trim()) params.set('trade_date', q.tradeDate.trim())
  if (q.session?.trim()) params.set('session', q.session.trim())
  if (q.strategyId?.trim()) params.set('strategy_id', q.strategyId.trim())
  if (q.accountId && q.accountId > 0) params.set('account_id', String(q.accountId))
  if (q.includeUserAction) params.set('include_user_action', 'true')

  const res = await fetch(`/api/opportunities/list?${params.toString()}`)
  const body = (await res.json()) as Record<string, unknown>
  if (!res.ok || !body?.ok) {
    throw new Error(String(body?.message || `HTTP ${res.status}`))
  }
  const pool = (body.pool || {}) as Record<string, unknown>
  const entriesRaw = Array.isArray(pool.entries) ? pool.entries : []
  return {
    accountId: Number(pool.account_id || pool.accountId) || 0,
    snapshotId: Number(pool.snapshot_id || pool.snapshotId) || 0,
    tradeDate: String(pool.trade_date || pool.tradeDate || ''),
    session: String(pool.session || ''),
    strategyId: String(pool.strategy_id || pool.strategyId || ''),
    scanBatchKey: String(pool.scan_batch_key || pool.scanBatchKey || ''),
    entries: entriesRaw.map((e) => mapEntry(e as Record<string, unknown>)),
    dataSourceNote: pool.data_source_note ? String(pool.data_source_note) : '',
  }
}

File: backend/api/opportunities.go
package api

import (
	"net/http"
	"strconv"
	"strings"

	"go-stock/backend/opportunity"
	"go-stock/backend/opportunity/outcome"
	"go-stock/backend/opportunity/projection"
)

// OpportunitiesHandler serves Phase14-G1.1 opportunity user action APIs.
//   POST /api/opportunities/action
//   GET  /api/opportunities/list
//   GET  /api/opportunities/projections
//   GET  /api/opportunities/outcomes
type OpportunitiesHandler struct {
	projections *projection.Service
	outcomes    *outcome.Service
}

func NewOpportunitiesHandler() *OpportunitiesHandler {
	return &OpportunitiesHandler{
		projections: projection.NewService(),
		outcomes:    outcome.NewService(),
	}
}

func (h *OpportunitiesHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	path := strings.TrimSuffix(r.URL.Path, "/")
	switch {
	case path == "/api/opportunities/action":
		h.handlePostAction(w, r)
	case path == "/api/opportunities/list":
		h.handleList(w, r)
	case path == "/api/opportunities/projections":
		h.handleProjections(w, r)
	case path == "/api/opportunities/outcomes":
		h.handleOutcomes(w, r)
	default:
		http.NotFound(w, r)
	}
}

type opportunityActionRequest struct {
	AccountID     uint   `json:"account_id"`
	ScanBatchKey  string `json:"scan_batch_key"`
	OpportunityID string `json:"opportunity_id"`
	StockCode     string `json:"stock_code"`
	Secucode      string `json:"secucode"`
	SignalTime    string `json:"signal_time"`
	SignalTag     string `json:"signal_tag"`
	Action        string `json:"action"`
	CreatedBy     string `json:"created_by"`
}

func (h *OpportunitiesHandler) handlePostAction(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeJSON(w, http.StatusMethodNotAllowed, map[string]any{
			"code": 405, "ok": false, "message": "POST required",
		})
		return
	}
	var req opportunityActionRequest
	if err := decodeJSONBody(r, &req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]any{
			"code": 400, "ok": false, "message": "invalid body: " + err.Error(),
		})
		return
	}

	row, err := opportunity.SaveUserOpportunityAction(opportunity.SaveUserOpportunityActionInput{
		AccountID:     req.AccountID,
		ScanBatchKey:  req.ScanBatchKey,
		OpportunityID: req.OpportunityID,
		StockCode:     req.StockCode,
		Secucode:      req.Secucode,
		SignalTime:    req.SignalTime,
		SignalTag:     req.SignalTag,
		Action:        req.Action,
		CreatedBy:     req.CreatedBy,
	})
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]any{
			"code": 400, "ok": false, "message": err.Error(),
		})
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"code":   0,
		"ok":     true,
		"action": row,
	})
}

func (h *OpportunitiesHandler) handleList(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeJSON(w, http.StatusMethodNotAllowed, map[string]any{
			"code": 405, "ok": false, "message": "GET required",
		})
		return
	}
	q := r.URL.Query()
	includeUser := strings.EqualFold(q.Get("include_user_action"), "true") ||
		q.Get("include_user_action") == "1"

	var accountID uint
	if raw := strings.TrimSpace(q.Get("account_id")); raw != "" {
		if v, err := strconv.ParseUint(raw, 10, 64); err == nil {
			accountID = uint(v)
		}
	}

	view, err := opportunity.ListFromSnapshot(opportunity.ListQuery{
		TradeDate:         strings.TrimSpace(q.Get("trade_date")),
		Session:           strings.TrimSpace(q.Get("session")),
		StrategyID:        strings.TrimSpace(q.Get("strategy_id")),
		AccountID:         accountID,
		IncludeUserAction: includeUser,
	})
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]any{
			"code": 500, "ok": false, "message": err.Error(),
		})
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"code": 0,
		"ok":   true,
		"pool": view,
	})
}

// RegisterOpportunitiesRoutes mounts opportunity routes (httptest).
func RegisterOpportunitiesRoutes(mux *http.ServeMux) {
	h := NewOpportunitiesHandler()
	mux.Handle("/api/opportunities/action", h)
	mux.Handle("/api/opportunities/list", h)
	mux.Handle("/api/opportunities/projections", h)
	mux.Handle("/api/opportunities/outcomes", h)
}

// OpportunitiesAssetMiddleware mounts /api/opportunities/* on Wails AssetServer.
func OpportunitiesAssetMiddleware(next http.Handler) http.Handler {
	h := NewOpportunitiesHandler()
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if strings.HasPrefix(r.URL.Path, "/api/opportunities/") {
			h.ServeHTTP(w, r)
			return
		}
		next.ServeHTTP(w, r)
	})
}
